File: dsa/9.stack_and_queue/1047_Remove_All_Adjacent_Duplicates_In_String.java
package stack_and_queue;
import java.util.*;

class Solution {
    public String removeDuplicates(String s) {
        Stack<Character> sStack = new Stack<Character>();
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < s.length(); i ++){
            if (sStack.isEmpty() == false && sStack.peek() == s.charAt(i)){
                while (sStack.isEmpty() == false && sStack.peek() == s.charAt(i)){
                    sStack.pop();
                }
            }
            else{
                sStack.push(s.charAt(i));
            }
        }

        while (sStack.isEmpty() == false){
            result.append(sStack.pop());
        }

        return result.reverse().toString();
    }
}

File: dsa/9.stack_and_queue/1047_Remove_All_Adjacent_Duplicates_In_String.py
import typing as t

class Solution:
    def removeDuplicates(self, s: str) -> str:
        stack: t.List[str] = []
        for char in s:
            if stack and char == stack[-1]:
                stack.pop()
            else:
                stack.append(char)
        
        return "".join(stack)
